tests/usm/usm_api.h

```diff
@@ -0,0 +1,688 @@
+ */
+constexpr auto multiple_events = 10_events;
+
+/** @brief Helper class providing number of elements for generators buffers if
+ *         numEvents > 0
+ */
+template <size_t numEvents>
+struct gen_buf_size_selector {
+  static constexpr size_t value = 10000;
+};
+
+/** @brief Helper class specialization providing number of elements for
+ *         generators buffers if numEvents == 0
+ */
+template <>
+struct gen_buf_size_selector<0> {
+  static constexpr size_t value = 0;
+};
+
+/** @brief Helper class for generation of 'sycl::event's by calling
+ *         queue::submit with long operation in kernel execution.
+ *  @tparam buf_size Size of arrays to process
+ */
+template <typename T, size_t buf_size>
+class event_generator {
+  sycl::range<1> rng{buf_size};
+  std::array<T, buf_size> arr_src;
+  std::array<T, buf_size> arr_dst;
+  sycl::buffer<T, 1> buf_src;
+  sycl::buffer<T, 1> buf_dst;
+
+ public:
+  event_generator()
+      : rng{buf_size},
+        buf_src{arr_src.data(), rng},
+        buf_dst{arr_dst.data(), rng} {
+    arr_src.fill(T{0});
+    arr_dst.fill(T{0});
+  }
+
+  /** @brief Initialize arr_src with init_value and return sycl::event of
+   *         queue::submit()
+   *  @param value Some non-zero (non-default) value of type T
+   */
+  sycl::event init(sycl::queue &queue, T value) {
+    return queue.submit([&](sycl::handler &cgh) {
+      auto acc_src = buf_src.template get_access<sycl::access_mode::write>(cgh);
+      // single_task is used to make process long enough for testing purpose
+      // The function being tested must wait for this task to complete.
+      cgh.single_task<init_kernel_name<T, buf_size>>([=]() {
+        for (size_t i = 0; i < buf_size; ++i) {
+          acc_src[i] = value;
+        }
+      });
+    });
+  }
+
+  /** @brief Copy data from arr_src to arr_dst for future check
+   */
+  void copy_arrays(sycl::queue &queue) {
+    queue.submit([&](sycl::handler &cgh) {
+      using kernel_name = copy_arrays_kernel_name<T, buf_size>;
+      auto acc_src = buf_src.template get_access<sycl::access_mode::read>(cgh);
+      auto acc_dst = buf_dst.template get_access<sycl::access_mode::write>(cgh);
+      // Copy should be much faster then algorithm in 'init()' method to detect
+      // situation when tested function doesn't wait for passed events
+      cgh.parallel_for<kernel_name>(rng, [=](sycl::id<1> idx) {
+        const size_t i = idx[0];
+        acc_dst[i] = acc_src[i];
+      });
+    });
+  }
+
+  /** @brief Check that elements of arrays are equal to each other and to
+   *         value
+   *  @param value The same value as value passed to init() method
+   */
+  bool check(T value) {
+    bool result = true;
+    auto acc_src = buf_src.template get_access<sycl::access_mode::read>(rng);
+    auto acc_dst = buf_dst.template get_access<sycl::access_mode::read>(rng);
+    for (size_t i = buf_size; i > 0; --i) {
+      const size_t idx{i - 1};
```

Comment:
Kind of cryptic to see the bounds
```suggestion
    for (size_t i = buf_size - 1; i + 1 > 0; --i) {
```

Reply:
I've applied this comment in this commit: 55155b57f5c5d74d0d7cd465d584da9f374f1a6e